vm/actor/src/builtin/puppet/mod.rs

```diff
@@ -0,0 +1,217 @@
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use num_derive::FromPrimitive;
+use num_traits::FromPrimitive;
+
+use crate::check_empty_params;
+use address::Address;
+use encoding::{tuple::*, Cbor};
+use ipld_blockstore::BlockStore;
+use num_bigint::bigint_ser;
+use runtime::{ActorCode, Runtime};
+use serde::de::{self, Deserializer};
+use serde::ser::{self, Serializer};
+use serde::{Deserialize, Serialize};
+
+use vm::{ActorError, ExitCode, MethodNum, Serialized, TokenAmount, METHOD_CONSTRUCTOR};
+
+// * Updated to specs-actors commit: e3ae346e69f7ad353b4eab6c20d8c6a5f497a039
+
+#[derive(FromPrimitive)]
+#[repr(u64)]
+pub enum Method {
+    Constructor = METHOD_CONSTRUCTOR,
+    Send = 2,
+    SendMarshalCBORFailure = 3,
+    ReturnMarshalCBORFailure = 4,
+    RuntimeTransactionMarshalCBORFailure = 5,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SendParams {
+    pub to: Address,
+    #[serde(with = "bigint_ser")]
+    pub value: TokenAmount,
+    pub method: u64,
+    pub params: Vec<u8>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SendReturn {
+    pub return_bytes: Serialized,
+    pub code: ExitCode,
+}
+
+#[derive(Default)]
+pub struct FailToMarshalCBOR {}
+
+impl Serialize for FailToMarshalCBOR {
+    fn serialize<S>(&self, _serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        Err(ser::Error::custom(
+            "Automatic fail when serializing FailToMarshalCBOR",
+        ))
+    }
+}
+
+impl<'de> Deserialize<'de> for FailToMarshalCBOR {
+    fn deserialize<D>(_deserializer: D) -> Result<Self, <D as Deserializer<'de>>::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Err(de::Error::custom(
+            "Automatic fail when deserializing FailToMarshalCBOR",
+        ))
+    }
+}
+
+impl Cbor for FailToMarshalCBOR {}
+
+#[derive(Default, Serialize_tuple, Deserialize_tuple)]
+pub struct State {
+    opt_fail: Vec<Option<FailToMarshalCBOR>>,
+}
+
+impl Cbor for State {}
+
+pub struct Actor;
+
+impl Actor {
+    fn constructor<BS, RT>(rt: &mut RT) -> Result<(), ActorError>
+    where
+        BS: BlockStore,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_accept_any()?;
+
+        rt.create(&State::default())?;
+        Ok(())
+    }
+
+    fn send<BS, RT>(rt: &mut RT, params: SendParams) -> Result<SendReturn, ActorError>
+    where
+        BS: BlockStore,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_accept_any()?;
+
+        let res = rt.send(
+            params.to,
+            params.method,
+            Serialized::serialize(params.params).unwrap(),
+            params.value,
+        );
+
+        let return_bytes = res.map_err(|_| {
+            ActorError::new(
+                ExitCode::ErrIllegalState,
+                "failed to get return bytes".to_string(),
+            )
+        })?;
+
+        Ok(SendReturn {
+            return_bytes,
+            code: ExitCode::Ok,
+        })
```

Comment:
```suggestion
        match res {
            Ok(return_bytes) => Ok(SendReturn {
                return_bytes: Some(return_bytes),
                code: ExitCode::Ok,
            }),
            Err(e) => Ok(SendReturn {
                return_bytes: None,
                code: e.exit_code(),
            }),
        }
```
You also need to change the return_bytes field to be an option, the go type can (and is nil in cases where there was an error)